datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt

```diff
@@ -154,10 +232,76 @@ private val Questionnaire.itemContextNameToExpressionMap: Map<String, String>
       .toMap()
   }
 
+/**
+ * Extracts a list containing the resource name followed by field names leading to the destination
+ * field defined in the [definition] field, or `null` if the [definition] field is empty or invalid.
+ *
+ * For example, if the [definition] field is
+ * "http://hl7.org/fhir/StructureDefinition/Patient#Patient.name", `listOf("Patient", "name")` will
+ * be returned.
+ */
+private val Questionnaire.QuestionnaireItemComponent.targetResourceAndElement: List<String>?
+  get() {
+    val snapshotPath = definition.substringAfter('#', "")
+    if (!"[a-zA-Z]+(\\.[a-zA-Z]+)+".toRegex().matches(snapshotPath)) return null
+
+    return snapshotPath.split(".")
+  }
+
+/**
+ * Retrieves details about the target field defined in
+ * [org.hl7.fhir.r4.model.Questionnaire.QuestionnaireItemComponent.definition] for easier searching
+ * of the setter methods and converting the answer to the expected parameter type.
+ */
+private val Questionnaire.QuestionnaireItemComponent.inferPropertyResourceClass: Field?
+  get() {
+    val modelAndField = targetResourceAndElement ?: return null
+    var resourceClass: Class<*> = Class.forName("org.hl7.fhir.r4.model.${modelAndField.get(0)}")
+    var resourceField: Field? = null
+
+    modelAndField.forEachIndexed loop@{ index, fieldName ->
+      if (index == 0) return@loop
+
+      resourceField = resourceClass.getFieldOrNull(fieldName) ?: return null
+      resourceClass = resourceField!!.mainType
+    }
+
+    return resourceField
+  }
+
 /**
  * See
  * [Extension: item extraction context](http://build.fhir.org/ig/HL7/sdc/StructureDefinition-sdc-questionnaire-itemExtractionContext.html)
  * .
  */
 private const val ITEM_CONTEXT_EXTENSION_URL: String =
   "http://hl7.org/fhir/uv/sdc/StructureDefinition/sdc-questionnaire-itemContext"
+
+private val Field.isList: Boolean
+  get() = isParameterized && type == List::class.java
+
+private val Field.isParameterized: Boolean
+  get() = genericType is ParameterizedType
+
+/* The type of the parameter in the setter for this field. */
+private val Field.setterParamType: Class<*>
+  get() = type
+
+private fun Field.retrieveNonParameterizedType(): Class<*> =
+  (genericType as ParameterizedType).actualTypeArguments[0] as Class<*>
+
+/**
+ * The non-parameterized type in this field.
+ *
+ * For example a Field of List<String> should give [String] as the [mainType]
+ */
+private val Field.mainType: Class<*>
```

Comment:
what about a more specific name:

```suggestion
private val Field.nonParameterizedType: Class<*>
```

Reply:
Done